### PR TITLE
Move all pluggy specifics into the manager

### PR DESCRIPTION
**Pull request by kenodegard**
### Description

Recent plugin hooks have implemented helper methods in the CondaPluginManager class instead of implementing pluggy specific code throughout conda. This does the same for the solver and virtual packages.

### Checklist - did you ...

- [ ] ~Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?~
     
- [x] Add / update necessary tests?
- [ ] ~Add / update outdated documentation?~

### CODE REVIEW

File: conda/plugins/manager.py [hunk 1]
```diff
@@ -249,6 +252,24 @@ def get_subcommands(self) -> dict[str, CondaSubcommand]:
             for subcommand in self.get_hook_results("subcommands")
         }
 
+    def get_virtual_packages(self) -> tuple[CondaVirtualPackage, ...]:
```
Comment (jezdez) [inline conda/plugins/manager.py:255]: Could we add a test for this?

Comment (kenodegard) [reply conda/plugins/manager.py:255]: already did 😄 https://github.com/conda/conda/pull/13057/commits/279fedcaa546797af99a9dbd882c20289105beca

File: conda/plugins/manager.py [hunk 2]
```diff
@@ -168,6 +168,13 @@ def get_hook_results(self, name: str) -> list:
             )
         return plugins
 
+    def get_solvers(self) -> dict[str, type[Solver]]:
+        """Return a mapping from solver name to solver class."""
+        return {
+            solver.name.lower(): solver.backend
+            for solver in self.get_hook_results("solvers")
+        }
```
Comment (jezdez) [inline conda/plugins/manager.py:176]: The pattern I thought those getters were supposed to have didn't include deconstructing the domain object, but return a mapping between the name of the plugin instance and the domain object

```suggestion
        return {
            solver_hook.name.lower(): solver
            for solver in self.get_hook_results("solvers")
        }
```

File: conda/plugins/manager.py [hunk 3]
```diff
@@ -184,12 +191,7 @@ def get_solver_backend(self, name: str | None = None) -> type[Solver]:
...
 
-        # Build a mapping between a lower cased backend name and
-        # solver backend class provided by the installed plugins.
-        solvers_mapping = {
-            solver.name.lower(): solver.backend
-            for solver in self.get_hook_results("solvers")
-        }
+        solvers_mapping = self.get_solvers()
 
         # Look up the solver mapping and fail loudly if it can't
```
Comment (jezdez) [inline conda/plugins/manager.py:196]: It's worth including the error handling in the getter

**Review by jezdez — CHANGES_REQUESTED** [review]
This is good, we've learned more about how to structure these plugins, but let's be careful about having the plugin manager only handle the plugin domain objects (in types.py) and not the specific plugin backends in the getter methods

File: conda/plugins/manager.py (hunk 1 above)
Comment (jezdez) [reply conda/plugins/manager.py:255]: Like codecov below shows that some branches aren't covered :D

File: conda/plugins/manager.py (hunk 3 above)
Comment (kenodegard) [reply conda/plugins/manager.py:196]: what error handling are you referring to?

File: conda/plugins/manager.py (hunk 2 above)
Comment (kenodegard) [reply conda/plugins/manager.py]: see https://github.com/conda/conda/pull/13057/commits/86497e13084e52de7fcc7790fbd26345e7a3e1e2

File: conda/plugins/manager.py (hunk 1 above)
Comment (kenodegard) [reply conda/plugins/manager.py:305]: Ah i see what you mean

**Review by jezdez — APPROVED** [review]

File: conda/plugins/manager.py [hunk 4]
```diff
@@ -117,24 +126,51 @@ def load_entrypoints(self, group: str, name: str | None = None) -> int:
                     )
                     continue
 
-                # only use the canonical name after this point
-                canonical = self.get_canonical_name(plugin)
-
-                # skip plugin if already registered or blocked
-                if self.get_plugin(canonical) or self.is_blocked(canonical):
-                    continue
```
Comment (jaimergp) [inline conda/plugins/manager.py:125]: @kenodegard - if I restore this block then things work as expected again.

`conda-libmamba-solver` is for some reason loaded twice in the `defaults` Docker image (first and last):

```
EntryPoint(name='conda-libmamba-solver', value='conda_libmamba_solver.plugin', group='conda')
EntryPoint(name='collect-upstream-conda-tests', value='collect_upstream_conda_tests', group='pytest11')
EntryPoint(name='pyflakes', value='pyflakes.api:main', group='console_scripts')
EntryPoint(name='coverage', value='coverage.cmdline:main', group='console_scripts')
EntryPoint(name='coverage-3.9', value='coverage.cmdline:main', group='console_scripts')
EntryPoint(name='coverage3', value='coverage.cmdline:main', group='console_scripts')
EntryPoint(name='rerunfailures', value='pytest_rerunfailures', group='pytest11')
EntryPoint(name='pytest_forked', value='pytest_forked', group='pytest11')
EntryPoint(name='pycodestyle', value='pycodestyle:_main', group='console_scripts')
EntryPoint(name='conda-verify', value='conda_verify.cli:cli', group='console_scripts')
EntryPoint(name='cph', value='conda_package_handling.cli:main', group='console_scripts')
EntryPoint(name='nodeenv', value='nodeenv:main', group='console_scripts')
EntryPoint(name='timeout', value='pytest_timeout', group='pytest11')
EntryPoint(name='anaconda-project', value='anaconda_project.cli:main', group='console_scripts')
EntryPoint(name='alias', value='setuptools.command.alias:alias', group='distutils.commands')
EntryPoint(name='bdist_egg', value='setuptools.command.bdist_egg:bdist_egg', group='distutils.commands')
EntryPoint(name='bdist_rpm', value='setuptools.command.bdist_rpm:bdist_rpm', group='distutils.commands')
EntryPoint(name='build', value='setuptools.command.build:build', group='distutils.commands')
EntryPoint(name='build_clib', value='setuptools.command.build_clib:build_clib', group='distutils.commands')
EntryPoint(name='build_ext', value='setuptools.command.build_ext:build_ext', group='distutils.commands')
EntryPoint(name='build_py', value='setuptools.command.build_py:build_py', group='distutils.commands')
EntryPoint(name='develop', value='setuptools.command.develop:develop', group='distutils.commands')
EntryPoint(name='dist_info', value='setuptools.command.dist_info:dist_info', group='distutils.commands')
EntryPoint(name='easy_install', value='setuptools.command.easy_install:easy_install', group='distutils.commands')
EntryPoint(name='editable_wheel', value='setuptools.command.editable_wheel:editable_wheel', group='distutils.commands')
EntryPoint(name='egg_info', value='setuptools.command.egg_info:egg_info', group='distutils.commands')
EntryPoint(name='install', value='setuptools.command.install:install', group='distutils.commands')
EntryPoint(name='install_egg_info', value='setuptools.command.install_egg_info:install_egg_info', group='distutils.commands')
EntryPoint(name='install_lib', value='setuptools.command.install_lib:install_lib', group='distutils.commands')
EntryPoint(name='install_scripts', value='setuptools.command.install_scripts:install_scripts', group='distutils.commands')
EntryPoint(name='rotate', value='setuptools.command.rotate:rotate', group='distutils.commands')
EntryPoint(name='saveopts', value='setuptools.command.saveopts:saveopts', group='distutils.commands')
EntryPoint(name='sdist', value='setuptools.command.sdist:sdist', group='distutils.commands')
EntryPoint(name='setopt', value='setuptools.command.setopt:setopt', group='distutils.commands')
EntryPoint(name='test', value='setuptools.command.test:test', group='distutils.commands')
EntryPoint(name='upload_docs', value='setuptools.command.upload_docs:upload_docs', group='distutils.commands')
EntryPoint(name='dependency_links', value='setuptools.dist:assert_string_list', group='distutils.setup_keywords')
EntryPoint(name='eager_resources', value='setuptools.dist:assert_string_list', group='distutils.setup_keywords')
EntryPoint(name='entry_points', value='setuptools.dist:check_entry_points', group='distutils.setup_keywords')
EntryPoint(name='exclude_package_data', value='setuptools.dist:check_package_data', group='distutils.setup_keywords')
EntryPoint(name='extras_require', value='setuptools.dist:check_extras', group='distutils.setup_keywords')
EntryPoint(name='include_package_data', value='setuptools.dist:assert_bool', group='distutils.setup_keywords')
EntryPoint(name='install_requires', value='setuptools.dist:check_requirements', group='distutils.setup_keywords')
EntryPoint(name='namespace_packages', value='setuptools.dist:check_nsp', group='distutils.setup_keywords')
EntryPoint(name='package_data', value='setuptools.dist:check_package_data', group='distutils.setup_keywords')
EntryPoint(name='packages', value='setuptools.dist:check_packages', group='distutils.setup_keywords')
EntryPoint(name='python_requires', value='setuptools.dist:check_specifier', group='distutils.setup_keywords')
EntryPoint(name='setup_requires', value='setuptools.dist:check_requirements', group='distutils.setup_keywords')
EntryPoint(name='test_loader', value='setuptools.dist:check_importable', group='distutils.setup_keywords')
EntryPoint(name='test_runner', value='setuptools.dist:check_importable', group='distutils.setup_keywords')
EntryPoint(name='test_suite', value='setuptools.dist:check_test_suite', group='distutils.setup_keywords')
EntryPoint(name='tests_require', value='setuptools.dist:check_requirements', group='distutils.setup_keywords')
EntryPoint(name='use_2to3', value='setuptools.dist:invalid_unless_false', group='distutils.setup_keywords')
EntryPoint(name='zip_safe', value='setuptools.dist:assert_bool', group='distutils.setup_keywords')
EntryPoint(name='PKG-INFO', value='setuptools.command.egg_info:write_pkg_info', group='egg_info.writers')
EntryPoint(name='dependency_links.txt', value='setuptools.command.egg_info:overwrite_arg', group='egg_info.writers')
EntryPoint(name='eager_resources.txt', value='setuptools.command.egg_info:overwrite_arg', group='egg_info.writers')
EntryPoint(name='entry_points.txt', value='setuptools.command.egg_info:write_entries', group='egg_info.writers')
EntryPoint(name='namespace_packages.txt', value='setuptools.command.egg_info:overwrite_arg', group='egg_info.writers')
EntryPoint(name='requires.txt', value='setuptools.command.egg_info:write_requirements', group='egg_info.writers')
EntryPoint(name='top_level.txt', value='setuptools.command.egg_info:write_toplevel_names', group='egg_info.writers')
EntryPoint(name='keywords', value='setuptools.dist:Distribution._finalize_setup_keywords', group='setuptools.finalize_distribution_options')
EntryPoint(name='parent_finalize', value='setuptools.dist:_Distribution.finalize_options', group='setuptools.finalize_distribution_options')
EntryPoint(name='cffi_modules', value='cffi.setuptools_ext:cffi_modules', group='distutils.setup_keywords')
EntryPoint(name='conda-content-trust', value='conda_content_trust.cli:cli', group='console_scripts')
EntryPoint(name='pip', value='pip._internal.cli.main:main', group='console_scripts')
EntryPoint(name='pip3', value='pip._internal.cli.main:main', group='console_scripts')
EntryPoint(name='pip3.9', value='pip._internal.cli.main:main', group='console_scripts')
EntryPoint(name='conda-pack', value='conda_pack.cli:main', group='console_scripts')
EntryPoint(name='conda-libmamba-solver', value='conda_libmamba_solver.plugin', group='conda')
```

Comment (jezdez) [reply conda/plugins/manager.py:125]: Oooh, interesting, why is the solver loaded twice, though? In doubt, I'd suggest to ignore the second run-through again and undo this part

Comment (jezdez) [reply conda/plugins/manager.py:125]: I've opened https://github.com/conda/conda/pull/13118 for that

**Pull request merged**